### PR TITLE
Text spotter postprocess

### PR DESCRIPTION
**Pull request by Harold-lkk**
## Motivation
Design a BasePostProcessor for text spotter

### CODE REVIEW

**Review by gaotongxiao — COMMENTED** [review]
Some comments

File: mmocr/models/textdet/postprocess/abcnet_postprocessor.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    # @force_fp32(apply_to=('cls_scores', 'bbox_preds'))
+
+    def filter_and_location(self,
+                            det_results,
+                            img_meta=None,
+                            nms_pre=-1,
+                            score_thr=0,
+                            max_per_img=100,
+                            nms=dict(type='IoULoss', loss_weight=1.0),
```
Comment (gaotongxiao) [inline mmocr/models/textdet/postprocess/abcnet_postprocessor.py:47]:
```suggestion
                            nms_cfg=dict(iou_thr=0.5),
```
We may need `train_cfg` and `test_cfg` to specify different threshold values

**Review by gaotongxiao — COMMENTED** [review]
A few other comments

File: mmocr/models/textdet/postprocess/abcnet_postprocessor.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        parameters = dict(
+            img_shape=img_meta['img_shape'],
+            nms_pre=nms_pre,
+            score_thr=score_thr)
+        (mlvl_bboxes, mlvl_scores, mlvl_labels, mlvl_score_factors,
+         mlvl_beziers) = multi_apply(self._single_level, cls_scores,
+                                     bbox_preds, centerness_preds,
+                                     bezier_preds, mlvl_priors, parameters)
+
+        mlvl_bboxes = torch.cat(mlvl_bboxes)
```
[the remark this answers is not in the document]
Comment (Harold-lkk) [reply mmocr/models/textdet/postprocess/abcnet_postprocessor.py:64]: don't need

File: mmocr/models/textdet/postprocess/base_postprocessor.py [hunk 3]
```diff
@@ -13,3 +19,106 @@ def is_valid_instance(self, area, confidence, area_thresh,
...
+    def reconstruct_text_instance(self, results, **kwargs):
+        return results
+
+    def split_results(self, pred_results, img_metas, **kwargs):
+        """convert pred_results to the follow format:
+
+        list(dict()) the list' size is batch size dict contain single image
+        pred result
+        """
+        return pred_results
```
[the remark this answers is not in the document]
Comment (Harold-lkk) [reply mmocr/models/textdet/postprocess/base_postprocessor.py:122]: Defaults should be direct pass if results is meeting the requirement ?

File: mmocr/models/textdet/postprocess/base_postprocessor.py [hunk 4]
```diff
@@ -13,3 +19,106 @@ def is_valid_instance(self, area, confidence, area_thresh,
...
+            results = self.rescale_results(results,
+                                           img_meta[0]['scale_factor'],
+                                           property)
+
+        if rescale_extra_property and extra_property is not None:
+            for key in extra_property:
+                assert key in results
+            results = self.rescale_results(results,
+                                           img_meta[0]['scale_factor'],
+                                           extra_property)
```
[the remark this answers is not in the document]
Comment (Harold-lkk) [reply mmocr/models/textdet/postprocess/base_postprocessor.py]:
```python
rescale_fields=dict(property=property, 
                     extra_property=extra_property,
                     rescale=rescale,
                     rescale_extra_property=rescale_extra_property)
``` 
 use parameters like this?

File: mmocr/models/textdet/postprocess/abcnet_postprocessor.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        self.bbox_coder = build_bbox_coder(bbox_coder)
+        self.use_sigmoid_cls = use_sigmoid_cls
+        if self.use_sigmoid_cls:
+            self.cls_out_channels = num_classes
+        else:
+            self.cls_out_channels = num_classes + 1
+
+    # @force_fp32(apply_to=('cls_scores', 'bbox_preds'))
+
+    def filter_and_location(self,
```
Comment (gaotongxiao) [inline mmocr/models/textdet/postprocess/abcnet_postprocessor.py:41]: If we finally decide to merge `filter_and_location` and `reconstruct_text_instance`, a proper name can be `get_text_instances`...

File: mmocr/models/textdet/postprocess/base_postprocessor.py [hunk 6]
```diff
@@ -13,3 +20,103 @@ def is_valid_instance(self, area, confidence, area_thresh,
                           confidence_thresh):
 
         return bool(area >= area_thresh and confidence > confidence_thresh)
+
+
+class BaseTextDetPostProcessor(nn.Module):
```
Comment (gaotongxiao) [inline mmocr/models/textdet/postprocess/base_postprocessor.py:25]:
```suggestion
class BaseDetPostprocessor(nn.Module):
```

File: mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from functools import partial
+
+import numpy as np
+import torch.nn as nn
+
+from mmocr.models.builder import build_postprocessor
+
+
+class BaseTwoStageTextSpotterPostProcessor(nn.Module):
```
Comment (gaotongxiao) [inline mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py:10]: too long

File: mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                **kwargs):
+        cfg = self.train_cfg if self.training else self.test_cfg
+        cfg.update(kwargs)
+        if self.det_postprocessor is not None:
+            det_results = self.det_postprocessor(
+                det_pred_results=det_pred_results,
+                img_metas=img_metas,
+                **kwargs)
+        else:
+            assert type(det_pred_results) is list
```
Comment (gaotongxiao) [inline mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py:47]:
```suggestion
            assert isinstance(det_pred_results, list)
```

File: mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            det_results = det_pred_results
+        if self.recog_postprocessor is not None:
+            recog_results = self.recog_postprocessor(recog_pred_results)
+        else:
+            assert type(recog_pred_results) is list
```
Comment (gaotongxiao) [inline mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py:52]:
```suggestion
            assert isinstance(recog_pred_results, list)
```

File: mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _rescale_single_result(self, polygon, scale_factor):
+        point_num = len(polygon)
+        assert point_num % 2 == 0
+        polygon = (np.array(polygon) *
+                   (np.tile(scale_factor[:2], int(point_num / 2)).reshape(
+                       1, -1))).flatten().tolist()
+        return polygon
+
+    def merge_text_spotter_result(self, det_result, recog_results):
```
Comment (gaotongxiao) [inline mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py:83]:
```suggestion
    def merge_text_spotter_result(self, det_results, recog_results):
```

File: mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        self.test_cfg = test_cfg
+        if det_postprocessor:
+            det_train_cfg = train_cfg.pop('det_postprocessor', None)
+            det_test_cfg = test_cfg.pop('det_postprocessor', None)
+
+            det_postprocessor.update(
+                dict(train_cfg=det_train_cfg, test_cfg=det_test_cfg))
+            self.det_postprocessor = build_postprocessor(det_postprocessor)
+        else:
+            self.det_postprocessor = None
```
Comment (gaotongxiao) [inline mmocr/models/textspotter/postprocess/base_two_stage_textspotter_postprocessor.py:27]: Make None as its default value to avoid the "else" branch. Same below

**Review by gaotongxiao — COMMENTED** [review]
LGTM

File: mmocr/models/textdet/postprocess/base_postprocessor.py [hunk 12]
```diff
@@ -13,3 +20,103 @@ def is_valid_instance(self, area, confidence, area_thresh,
...
+        assert text_repr_type in ['poly', 'quad']
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+    def forward(self, pred_results, img_metas=None, **kwargs):
+
+        cfg = self.train_cfg if self.training else self.test_cfg
+        cfg.update(kwargs)
+        if len(img_metas) == 1:
+            img_metas = [img_metas]
```
[the remark this answers is not in the document]
Comment (Harold-lkk) [reply mmocr/models/textdet/postprocess/base_postprocessor.py:49]: Yeah, img_metas is a 2d list when batch_size > 1, so I keep the same dim when batch size = 1.

File: mmocr/models/textdet/postprocess/abcnet_postprocessor.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+                labels=mlvl_labels.detach().cpu().numpy(),
+                bezier=mlvl_beziers.detach().cpu().numpy())
+            return results
+        det_bboxes, keep_idxs = batched_nms(mlvl_bboxes, mlvl_scores,
+                                            mlvl_labels, nms)
+        det_bboxes, scores = np.split(det_bboxes, [-1], axis=1)
+        results = dict(
+            bboxes=det_bboxes[:max_per_img].detach().cpu().numpy(),
+            scores=scores[:max_per_img].detach().cpu().numpy(),
+            labels=mlvl_labels[keep_idxs][:max_per_img].detach().cpu().numpy(),
```
Comment (gaotongxiao) [inline mmocr/models/textdet/postprocess/abcnet_postprocessor.py:120]: We shouldn't have `labels` for text detectors' outputs
